db/db_impl_compaction_flush.cc

```diff
@@ -200,6 +202,172 @@ Status DBImpl::FlushMemTableToOutputFile(
+        false /* sync_output_directory */, false /* write_manifest */);
+  }
+
+  autovector<FileMetaData> file_meta;
+  Status s;
+  // TODO (yanqin): parallelize jobs with threads.
+  for (int i = 0; i != num_cfs; ++i) {
+    auto& job = jobs[i];
+    job.PickMemTable();
+    file_meta.emplace_back(FileMetaData());
+
+#ifndef ROCKSDB_LITE
+    // may temporarily unlock and lock the mutex.
+    NotifyOnFlushBegin(cfds[i], &file_meta[i], mutable_cf_options_list[i],
+                       job_context->job_id, job.GetTableProperties());
+#endif /* !ROCKSDB_LITE */
+    if (logfile_number_ > 0 &&
+        versions_->GetColumnFamilySet()->NumberOfColumnFamilies() > 0) {
+      s = SyncClosedLogs(job_context);
+    }
+    if (!s.ok()) {
+      break;
+    }
+    s = job.Run(&logs_with_prep_tracker_, &file_meta[i]);
+    if (!s.ok()) {
+      break;
+    }
+  }
+
+  if (s.ok()) {
+    // Sync on all distinct output directories.
+    for (auto dir : distinct_output_dirs) {
+      if (dir != nullptr) {
+        dir->Fsync();
+      }
+    }
+
+    autovector<const autovector<MemTable*>*> mems_list;
+    autovector<MemTableList*> imm_lists;
+    for (int i = 0; i != num_cfs; ++i) {
+      const auto& mems = jobs[i].GetMemTables();
+      mems_list.emplace_back(&mems);
+      imm_lists.emplace_back(cfds[i]->imm());
+    }
+    s = MemTableList::InstallMemtableFlushResults(
+        imm_lists, cfds, mutable_cf_options_list, mems_list,
+        &logs_with_prep_tracker_, versions_.get(), &mutex_, file_meta,
+        &job_context->memtables_to_free, directories_.GetDbDir(), log_buffer);
+  }
+
+  if (s.ok()) {
+    for (int i = 0; i != num_cfs; ++i) {
+      InstallSuperVersionAndScheduleWork(cfds[i],
+                                         &job_context->superversion_contexts[i],
+                                         mutable_cf_options_list[i]);
+      VersionStorageInfo::LevelSummaryStorage tmp;
+      ROCKS_LOG_BUFFER(log_buffer, "[%s] Level summary: %s\n",
+                       cfds[i]->GetName().c_str(),
+                       cfds[i]->current()->storage_info()->LevelSummary(&tmp));
+    }
+    if (made_progress) {
+      *made_progress = 1;
+    }
+#ifndef ROCKSDB_LITE
+    auto sfm = static_cast<SstFileManagerImpl*>(
+        immutable_db_options_.sst_file_manager.get());
+    for (int i = 0; i != num_cfs; ++i) {
+      NotifyOnFlushCompleted(cfds[i], &file_meta[i], mutable_cf_options_list[i],
+                             job_context->job_id, jobs[i].GetTableProperties());
+      if (sfm) {
+        std::string file_path = MakeTableFileName(
+            cfds[i]->ioptions()->cf_paths[0].path, file_meta[i].fd.GetNumber());
+        sfm->OnAddFile(file_path);
+        if (sfm->IsMaxAllowedSpaceReached()) {
+          Status new_bg_error =
+              Status::SpaceLimit("Max allowed space was reached");
+          error_handler_.SetBGError(new_bg_error,
+                                    BackgroundErrorReason::kFlush);
+          break;
```

Comment:
People may rely on `OnFlushBegin` calls matching `OnFlushCompleted` calls, which won't happen if we break here.

Reply:
Never mind, I checked the existing code and it doesn't provide this guarantee. For some reason we try to provide it for files only...